### PR TITLE
Logging Overhaul

### PR DESCRIPTION
**Pull request by davemcorwin**
# Logging Overhaul
Herein we significantly change how logging is handled in this application. All logging-related logic is handled by leveraging Python's provided logging architecture. Logging performed by the application itself AND `stdout` and `stderr` streams from "task"s are sent through the logging architecture.

## Implications
- Logs are written and/or sent in realtime, no need to do this explicitly after a "task" is complete. Theoretically, this means we won't lose (all) logs in the case of an unexpected exit and we can share them with users in closer to real-time.
- All filtering (ie removing secrets) is handled at log-time by the loggers, via the `LogFilter`, no need for an explicit step
- Logging inside a "task" can just be done with `print` and no knowledge or use of the logging architecture, since the output streams are already mapped to loggers by the calling process.
- Because of ^, invoking one from the command line will NOT use the logging architecture described here, the output will just be sent to stdout and stderr. 
- Default data, such as the build id, repo, etc, ... is added to ALL log messages by default and configured in one location
- Multiple log streams can be created for different consumers with different levels of logging, formatting, and filtering.

## Log Streams
Log streams have been implemented for two consumers. Both are set to a logging level of `INFO`.

### Internal
For our internal use (stdout, stderr -> Cloud.gov/Kibana). This is meant for more verbose logging and includes:
- Build Id
- Org
- Repo
- Branch

Example:
```
2019-12-20 20:25:30 INFO [MAIN] @buildId: 1234 @owner: 18F @repo: federalist-uswds-gatsby2 @branch: master @message: Running build for 18F/federalist-uswds-gatsby2/master
```

Adding the attributes to ALL messages will help us debug problems by being able to easily filter by those attributes in Kibana.

### Partner Facing
For sharing with our users in the Federalist Web Application. These are sent directly to the database with a source of "ALL" and omit the additional attributes present in the internal logs.

Example:
```
2019-12-31 20:17:55 INFO [run-federalist-script] > node scripts/postinstall.js
```

## Operator Changes
The environment variable `LOG_CALLBACK` will be removed from use and replaced with:
- `DB_URL`: the url to the postgres database to which the app will stream logs. In a deployed environment this is provided by Cloud Foundry.

## Current Caveats
- The one exception to this is the `main` method since it is the entrypoint. It has been removed as a "task" to make it clear it must be run as plain ol' python.
- Initial plans were to potentially set the internal log handler to `DEBUG`, however, this includes logs from `pyinvoke` itself which is not what we want. If we do want to create different logging levels, we can create one in between `DEBUG` and `INFO`. For now, both are set to `INFO`.

## TODO
- [x] Unit tests for additional logging related logic
- [x] E2E tests for all build types
  - [x] Node.js
  - [x] Hugo
  - [x] Jekyll
  - [x] Static
  - [-] From template (test in staging)
- [x] ~Release of changes to web application to support http logging~ no longer necessary, the web app will just show what is there.
- [x] Optional - Launch `main` task in production with a script to standardize logging
- [x] Bind database service to CF apps in `staging`
- [ ] Bind database service to CF apps in `production`

### CODE REVIEW

File: bin/migrate.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import os
+import psycopg2
+
+
+def migrate():
+    conn = psycopg2.connect(os.environ['DB_URL'])
```
Comment (amirbey) [inline bin/migrate.py]: if we bind the DATABASE ... i think this variable will be DATABASE_URL

https://docs.cloudfoundry.org/devguide/deploy-apps/environment-variable.html#DATABASE-URL

Comment (amirbey) [inline bin/migrate.py]:
```suggestion
    conn = psycopg2.connect(os.environ['DATABASE_URL'])
```

File: docker-compose.yml [hunk 2]
```diff
@@ -10,12 +10,15 @@ services:
...
       - echoserver
+      - db
     depends_on:
       - echoserver
+      - db
     environment:
       FEDERALIST_BUILDER_CALLBACK: http://echoserver:8989/builder
       STATUS_CALLBACK: http://echoserver:8989/status
       LOG_CALLBACK: http://echoserver:8989/log
+      DB_URL: postgresql://postgres@db/federalist
```
Comment (amirbey) [inline docker-compose.yml]:
```suggestion
      DATABASE_URL: postgresql://postgres@db/federalist
```

File: bin/migrate.py (hunk 1 above)
Comment (amirbey) [reply bin/migrate.py]: https://docs.cloudfoundry.org/devguide/deploy-apps/environment-variable.html#DATABASE-URL

File: docker-compose.yml (hunk 2 above)
Comment (amirbey) [reply docker-compose.yml]: https://docs.cloudfoundry.org/devguide/deploy-apps/environment-variable.html#DATABASE-URL

File: .env.sample [hunk 3]
```diff
@@ -11,6 +11,7 @@ BUCKET=
 # FEDERALIST_BUILDER_CALLBACK=
 # STATUS_CALLBACK=
 # LOG_CALLBACK=
+DB_URL=
```
Comment (amirbey) [inline .env.sample]:
```suggestion
DATABASE_URL=
```

Comment (amirbey) [reply .env.sample]: https://docs.cloudfoundry.org/devguide/deploy-apps/environment-variable.html#DATABASE-URL

File: tasks/publish.py [hunk 4]
```diff
@@ -23,21 +17,14 @@ def publish(ctx, base_url, site_prefix, bucket, cache_control,
...
+    print('Publishing to S3')
 
     start_time = datetime.now()
 
-    s3_client = boto3.client(
-        service_name='s3',
-        region_name=aws_region,
-        aws_access_key_id=access_key_id,
-        aws_secret_access_key=secret_access_key)
+    s3_client = boto3.client(service_name='s3', region_name=aws_region)
```
Comment (amirbey) [inline tasks/publish.py]: is it the case that the access_key_id and aws_secret_access_key no longer needed if the env variables exist?

File: log_utils/get_logger.py [hunk 5]
```diff
@@ -3,26 +3,111 @@
...
+
+    stream_handler = logging.StreamHandler(sys.stdout)
+    stream_handler.setFormatter(stream_formatter)
+    stream_handler.setLevel(log_level)
+    stream_handler.addFilter(log_filter)
+
+    handlers = [stream_handler]
+
+    if not skip_logging and db_url:
+        buildId = attrs['buildid']
```
Comment (amirbey) [inline log_utils/get_logger.py]: buildid .... should this be "buildId" or "build_id"?

File: tasks/publish.py (hunk 4 above)
Comment (davemcorwin) [reply tasks/publish.py]: yeah, they can either be in the environment, provided as arguments, or one of the other million ways to specify them...

https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html#guide-configuration

File: log_utils/get_logger.py (hunk 5 above)
Comment (davemcorwin) [reply log_utils/get_logger.py]: i'll make it `build_id` for consistency

**Comment by davemcorwin** [comment]
@amirbey This is ready for staging pending updates to `federalist` web app.

**Pull request merged**